src/main/java/com/mparticle/kits/SingularKit.java

```diff
@@ -0,0 +1,235 @@
+    }
+
+    @Override
+    public List<ReportingMessage> logError(String s, Map<String, String> map) {
+        return null;
+    }
+
+    @Override
+    public List<ReportingMessage> logException(Exception e, Map<String, String> map, String s) {
+        return null;
+    }
+
+    @Override
+    public List<ReportingMessage> logEvent(MPEvent mpEvent) {
+        List<ReportingMessage> messages = new LinkedList<ReportingMessage>();
+        String eventName = mpEvent.getEventName();
+        Map eventInfo = mpEvent.getInfo();
+        if (eventInfo != null) {
+            JSONObject params = new JSONObject(eventInfo);
+            eventStatus = Singular.event(eventName, params.toString());
+        } else {
+            eventStatus = Singular.event(eventName);
+        }
+        if (eventStatus) {
+            messages.add(ReportingMessage.fromEvent(this, mpEvent));
+        } else {
+            messages.add(new ReportingMessage(this, ReportingMessage.MessageType.ERROR, System.currentTimeMillis(), null));
+        }
+        return messages;
+    }
+
+    @Override
+    public List<ReportingMessage> logScreen(String s, Map<String, String> map) {
+        return null;
+    }
+
+    @Override
+    public void handleLink(String link) {
+        DeepLinkListener deepLinkListener = MParticle.getInstance().getDeepLinkListener();
+        if (deepLinkListener != null) {
+            DeepLinkResult deepLinkResult = new DeepLinkResult();
+            deepLinkResult.setServiceProviderId(MParticle.ServiceProviders.SINGULAR);
+            deepLinkResult.setLink(link);
+            deepLinkListener.onResult(deepLinkResult);
+        }
+    }
+
+    @Override
+    public boolean willHandlePushMessage(Intent intent) {
+        return false;
+    }
+
+    @Override
+    public void onPushMessageReceived(Context context, Intent intent) {
+
+    }
+
+    @Override
+    public boolean onPushRegistration(String deviceToken, String senderId) {
+        try {
+            switch (MPUtility.getAvailableInstanceId()) {
+                case GCM:
+                    Singular.setGCMDeviceToken(deviceToken);
+                    break;
+                case FCM:
+                    Singular.setFCMDeviceToken(deviceToken);
+                    break;
+            }
+        } catch (Exception unableToSetDeviceToken) {
+            return false;
+        }
+        return true;
+    }
+
+    @Override
+    public void setInstallReferrer(Intent intent) {
+        new SingularInstallReceiver().onReceive(getContext(), intent);
+    }
+
+
+    @Override
+    public List<ReportingMessage> logLtvIncrease(BigDecimal bigDecimal, BigDecimal bigDecimal1, String s, Map<String, String> map) {
+        return null;
+    }
+
+    @Override
+    public List<ReportingMessage> logEvent(CommerceEvent commerceEvent) {
+        List<ReportingMessage> messages = new LinkedList<ReportingMessage>();
+        if (commerceEvent.getProductAction().equals(Product.PURCHASE)) {
```

Comment:
@anismiles this needs to be an `else`, otherwise you're going to double-count `PURCHASE` events. I suppose that wouldn't necessarily be harmful as revenue won't be counted twice, but I want to make sure you *meant* to code it like this.